support-modules/acquisition-events/src/main/scala/com/gu/support/acquisitions/bigquery/BigQueryService.scala

```diff
@@ -28,43 +42,79 @@ class BigQueryService(config: BigQueryConfig) {
+  private def insertForPromise(
+      acquisitionDataRow: AcquisitionDataRow,
+  ): Promise[Either[String, Unit]] = {
+    val tableId = TableName.of(config.projectId, datasetName, tableName)
+    val rowContent: JSONObject = AcquisitionDataRowMapper.mapToTableRow(acquisitionDataRow)
+    SafeLogger.info(s"Attempting to append row ($rowContent) created from ($acquisitionDataRow)")
+    val streamWriter = JsonStreamWriter.newBuilder(tableId.toString(), bigQueryWriteClient).build();
+    val promise = Promise[Either[String, Unit]]()
     try {
-      val tableId = TableId.of(datasetName, tableName)
-      val rowContent = AcquisitionDataRowMapper.mapToTableRow(acquisitionDataRow)
-      val insertRequest = InsertAllRequest.newBuilder(tableId).addRow(rowContent).build
-
-      val response = bigQuery.insertAll(insertRequest)
+      val responseFuture = streamWriter.append(new JSONArray(List(rowContent).asJava));
+      ApiFutures.addCallback(
+        responseFuture,
+        new BigQueryService.AppendCompleteCallback(promise),
+        // I reckon directExecutor is fine here because the callback just
+        // completes the promise
+        MoreExecutors.directExecutor(),
+      )
+    } catch {
+      case e: AppendSerializtionError =>
+        val errorMessage =
+          e.getRowIndexToErrorMessage().asScala.map { case (i, message) => s"$i: $message" }.mkString(", ")
+        SafeLogger.error(scrub"There was an exception appending to $tableName: $errorMessage", e)
+        promise.success(Left(s"Error appending to table $tableName: $errorMessage"))
+    }
+    promise
+  }
+}
 
-      if (response.hasErrors) {
-        val errors = response.getInsertErrors.entrySet.asScala.mkString(", ").stripSuffix(", ")
-        SafeLogger.error(scrub"Failed to insert row into $tableName: $errors")
-        Left(s"Failed to insert row into $tableName: $errors")
-      } else {
-        SafeLogger.info(s"Rows successfully inserted into table $tableName")
-        Right(())
+object BigQueryService {
+  case class AppendCompleteCallback(promise: Promise[Either[String, Unit]])
+      extends ApiFutureCallback[AppendRowsResponse] {
+    def onSuccess(response: AppendRowsResponse): Unit = {
+      SafeLogger.info(s"Rows successfully inserted into table $tableName")
+      promise.success(Right(()))
+    }
+    def onFailure(throwable: Throwable): Unit = {
+      val detail: String = throwable match {
+        case e: AppendSerializtionError => {
+          e.getRowIndexToErrorMessage().asScala.map { case (i, message) => s"$i: $message" }.mkString(", ")
+        }
+        case _ => ""
       }
-    } catch {
-      case e: BigQueryException =>
-        SafeLogger.error(scrub"There was an exception inserting a row into $tableName", e)
-        Left(s"There was an exception inserting a row into $tableName: ${e.getMessage}")
+      SafeLogger.error(scrub"There was an exception inserting a row into $tableName: $detail", throwable)
```

Comment:
This sounds very sensible – I’ll look into it. Pushing a metric to cloudwatch in the callback will probably mean replacing the direct executor with one that uses another thread somehow, which does sound a little daunting.

Reply:
I’ve made these changes now, using a single thread executor that’s created for the callback and shut down at the end of the callback. This seems ok, but I wonder whether it would be better to create a larger thread pool for each BigQueryService, rather than a small one for each callback? If I did that I’m not sure where/when we would shut down the thread pool though – I guess we would need to explicitly shut down BigQueryServices at each callsite.